mailet/api/src/test/java/org/apache/mailet/ContractMailTest.java

```diff
@@ -0,0 +1,133 @@
+
+package org.apache.mailet;
+
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+
+import com.google.common.collect.ImmutableMap;
+
+public abstract class ContractMailTest {
+    public final static AttributeName ATTRIBUTE_NAME_1 = AttributeName.of("name1"); 
+    public final static AttributeName ATTRIBUTE_NAME_2 = AttributeName.of("name2"); 
+    public final static Attribute ATTRIBUTE_1 = new Attribute(ATTRIBUTE_NAME_1, AttributeValue.of(true)); 
+    public final static Attribute ATTRIBUTE_2 = new Attribute(ATTRIBUTE_NAME_2, AttributeValue.of("value2")); 
+    public final static Attribute ATTRIBUTE_1_BIS = new Attribute(ATTRIBUTE_NAME_1, AttributeValue.of("value1")); 
+
+    public abstract Mail newMail();
+
+    @Nested
+    public class EmptyMail {
+        private Mail mail;
+        
+        @BeforeEach
+        public void setUp() {
+            mail = newMail();
+        }
+        
+        public void newMailShouldHaveNoAttributes() {
+            assertThat(mail.attributes()).isEmpty();
+        }
+        
+        @Test
+        public void newMailShouldHaveNoAttributeNames() {
+            assertThat(mail.attributeNames()).isEmpty();
+        }
+
+        @Test
+        public void newMailShouldHaveAnEmptyAttributeMap() {
+            assertThat(mail.attributesMap()).isEmpty();
+        }
+    }
+    
+    @Nested
+    public class OneAttributeMail {
+        private Mail mail;
+        
+        @BeforeEach
+        public void setUp() {
+            mail = newMail();
+            assertThat(mail.setAttribute(ATTRIBUTE_1)).isNull();
+        }
+        
+        @Test
+        public void shouldHaveOneAttribute() {
+            assertThat(mail.attributes()).containsExactly(ATTRIBUTE_1);
+        }
+        
+        @Test
+        public void shouldHaveOneAttributeName() {
+            assertThat(mail.attributeNames()).containsExactly(ATTRIBUTE_NAME_1);
+        }
+
+        @Test
+        public void shouldHaveOneAttributeMap() {
+            assertThat(mail.attributesMap()).isEqualTo(ImmutableMap.of(ATTRIBUTE_NAME_1, ATTRIBUTE_1));
+        }
+        
+        @Test
+        public void shouldBeRetrievable() {
+            assertThat(mail.getAttribute(ATTRIBUTE_NAME_1)).contains(ATTRIBUTE_1);
+        }
+        
+        @Test
+        public void shouldBeRemovable() {
+            assertThat(mail.removeAttribute(ATTRIBUTE_NAME_1)).isEqualTo(ATTRIBUTE_1);
+            assertThat(mail.getAttribute(ATTRIBUTE_NAME_1)).isEmpty();
+        }
+        
+        @Test
+        public void shouldBeReplacable() {
+            assertThat(mail.setAttribute(ATTRIBUTE_1_BIS)).isEqualTo(ATTRIBUTE_1);
+            assertThat(mail.getAttribute(ATTRIBUTE_NAME_1)).contains(ATTRIBUTE_1_BIS);
+        }
+    }
+    
+    @Nested
+    public class TwoAttributesMail {
```

Comment:
this class is missing mail.setAttribute(), mail.getAttribute(), mail.removeAttribute() tests. And my idea is naming it to `MultipleAttributesMail` where you can freely play tests with various number of attributes, not only 2. Could gain more confident.